### ISSUE TITLE
Scala3 parser support

### ISSUE THREAD

**Issue by kpbochenek** [scalameta/scalameta#2022]
As talking with @tgodzik this issue will cover work in regard to providing support in parser for scala3

Scala3 features:
- [x] Intersection / Union types `(A & B)` `(A | B)`
- [x] Type Lambda (`[X, Y] =>> Map[Y, X]`)
- [x] Given/Using definition
- [x] Given imports (`import a.{given _}`)
- [x] Using call side (`val a = f(3)(using intOrd)(using 3, 'c')`)
- [x] Enums (`enum Color { case R, G, B }`)
- [x] Extension methods/blocks (`extension x { ... }`)
- [x] inline
- [x] macros
- [x] extending multiple traits new syntax (`trait C extends A, B`)
- [x] trait parameters (`trait X(val i: Int`)
- [x] export clause (`export a.b.{ ... }`)
- [x] super trait
- [x] opaque type (`opaque type Logarithm = Double`)
- [x] open classes (`open class Writer[T] { ... }`)
- [x] no if/for parenthesis (`if x < 0 then "neg" else if x == 0 "zero" else "pos"`)
- [x] Optional braces
- [x] type wildcard (`List[_] -> List[?]`)
- [x] varargs pattern change (`case List(1, _ : _*) =>`
- [x] named type arguments (`f[A = Int]()`)  (fixes #903)
- [x] transparent inline methods
- [x] end marker
- [x] `as` for variable binding in pattern match
- [x] inline if (`inline if (a > b) a else b`)
- [x] polymorphic function types (`val a: [X] => Y => Z`) (https://github.com/lampepfl/dotty/blob/master/tests/run/polymorphic-functions.scala)
- [x] statement `type T` 
- [x] Match Types (`type List[X] = X match { ... }`)
- [x] use type for given imports `given [ Context[_] ]`
- [x] changed given imports `import a.b.c.given`
- [x] Derivation (`class X derives Eql`)
- [x] Context Functions (`type Executable[T] = ExecutionContext ?=> T`)
- [x] For comprehension new parenthesisless syntax
- [x] Match chaining and `.match` syntax
- [x] string interpolation pattern matching with capital letters
- [x] check quoted pattern - https://github.com/lampepfl/dotty/pull/10125
- [x] inline match (waiting on resolution of https://github.com/lampepfl/dotty/issues/10374)
- [x] infix as soft keyword for types and methods
- [x] dependent function types `val extractor: (e: Entry) => e.Key = extractKey` #2196
- [x] enum case should not be allowe to be outside enum #2194 
- [x] reworked given syntax with `:` instead of `as` 
- [x] `with` can start indentation for givens
- [x] dropped `as` soft keyword
- [x] extension group no longer requires `:` in an indentation region https://github.com/scalameta/scalameta/pull/2237
- [x] ~~erased modifier which will be added in 3.1~~ everything can really change so no point including it here
- [x] drop `super trait` (https://github.com/lampepfl/dotty/commit/67017d72719a4b780caf4ad4e27e9a24aaab89a0)
- [x] changed vararg splices https://dotty.epfl.ch/docs/reference/changed-features/vararg-splices.html https://github.com/scalameta/scalameta/pull/2238
- [x] `=>` replaced with `as` for import renames https://github.com/scalameta/scalameta/pull/2239
- [x] import all with `*` https://github.com/scalameta/scalameta/pull/2239
- [x] changed extensions to contain mutilple using and allowed for type params in defs https://github.com/scalameta/scalameta/pull/2242
- [x] observe outdent on `else`, `then`,  `do`,  `catch`, `finally`,  `yield` and  `match` on the same indentation level https://github.com/scalameta/scalameta/pull/2244

Deprecations:
- [x] drop given imports
- [x] probably remove anonymous given with at depending how https://github.com/lampepfl/dotty/issues/10074 is resolved
- [x] XML literals
- [x] drop support for `do ... while ...` https://github.com/scalameta/scalameta/pull/2314
- [x] drop procedure syntax https://github.com/scalameta/scalameta/pull/2313
- [x] ~~drop package object~~  this is still allowed, we will need to drop it for 3.1
- [x] ~~drop class shadowing~~ (@tgodzik I don't think this is something we can solve in the parser)
- [x] ~~drop unsupported vararg splices~~ https://dotty.epfl.ch/docs/reference/changed-features/vararg-splices.html this is still allowed, we will need to drop it for 3.1
- [x] ~~drop wildcards according to  https://dotty.epfl.ch/docs/reference/dropped-features/wildcard-init.html~~ this is still allowed, we will need to drop it for 3.1

Optional:

- [x] support `-Yindent-colons` - this is an experimental feature that is currently only avialable in nightlies, but should now be possible to implement when needed, but I would only work on it once it makes it way to the mainline Scala 3.

Other known work:
- [x] Update test `"error on vararg parameter by name"` to verify error is NOT thrown for dotty (#2126)
- [x] Check if OR types conflicting with or used in case match:
 // pattern match `case _: A | B => ...` should be equal to `case (_: A) | B => ...`
  // check if we create case with (_: A) | B or WRONGLY _: (A | B)
- [x] Adjust to changes (https://github.com/lampepfl/dotty/pull/9055/files)
- [x] Update code around comment '//NOTE(kpbochenek): Needs to be changed to ...'
- [x] Test: "indent-below-not-okay" discussion in https://github.com/lampepfl/dotty/issues/9790
- [x] Handle comment: "// Dotty related exceptions - To be removed upon releasing 4.3.25 or 4.4"

**Comment by kpbochenek** [comment]
[Question 1]
In dotty you no longer provide implicit arguments as normal arguments(according to Odersky it was confusing) and you have special syntax for that.
```scala
def f(s: String)(using Int) = ???   // definition
f("123")  // call this function and proper implicit Int implicit needs to be in scope
f("123")(using MyInt)  // explicitly providing implicit
f("123")(using 3)    // nothing new, we can provide literal, method call etc as parameter
```
Currently method call is translated e.g.:
```
f("123")(px, 3)   // purposefully 2 implicit params to show complex scenario
==>
Term.Apply(
  Term.Apply(
    Term.Name("f"),
    List(Lit.String("123"))
  ),
  List(Term.Name("px"), Term.Lit(3))
)
```

What would be best representation of this? :
```
f("123")(using px, 3)
==>
Term.Apply(
  Term.Apply(
    Term.Name("f"),
    List(Lit.String("123"))
  ),
  List(???)
)
```

**Comment by tgodzik** [comment]
I would do something like:
```
Term.Apply(
  Term.Apply(
    Term.Name("f"),
    List(Lit.String("123")),
    usingApply = false
  ),
  List(Term.Name("px"), Term.Lit(3)),
   usingApply = true
)
```

**Comment by olafurpg** [comment]
One design principle in `scala.meta.Tree` is that we don't use default parameter values so we should avoid `Term.Apply(.., usingApply = true)`. An alternative solution would be to do `Term.Using(term: Term)`

**Comment by kpbochenek** [comment]
Another problem: https://dotty.epfl.ch/docs/reference/contextual/derivation.html
Dotty allows to write `trait Tree extends A derives Eql, Show`

In our current structure:
```scala
@ast class Trait(mods: List[Mod],name: Type.Name,tparams: List[Type.Param],ctor: Ctor.Primary,templ: Template)
@ast class Template(early: List[Stat], inits: List[Init], self: Self, stats: List[Stat])
@ast class Init(tpe: Type, name: Name, argss: List[List[Term]])
```

I need to change `Template` to be:
```scala
@ast class Template(early: List[Stat], inits: List[Init], derives: List[Term.Name], self: Self, stats: List[Stat])
```

This will be breaking change but I can see no other way around it, maybe you have any plan how to make it backward compatible?
/cc @tgodzik @olafurpg

**Comment by olafurpg** [comment]
@kpbochenek I recall some discussions in the past about `Template` being a mistake and we should have inlined the fields into `Defn.{Object,Trait,Class}` instead. Is `derives` valid for all of trait/object/class?

**Comment by kpbochenek** [comment]
and `enum`, yes

**Comment by tgodzik** [comment]
Maybe adding a DerivesTemplate if the language supports `derives`?

**Comment by tgodzik** [comment]
Or have derives as a private value accessible via method?

**Comment by olafurpg** [comment]
Then I don't have a good answer except maybe to add `var derives: List[Term.Name]` to `Template` 🤷 We could make the mutation private to the `scala.meta` package

It would be good to have an estimate for how many other similar breaking changes we will need to make in order to support Scala 3. I'm open to breaking changes if we can do it in one round but it's problematic if we need several incremental iterations that each introduce breaking changes.

**Comment by kpbochenek** [comment]
There is still open question of how to represent significant indentation vs braces and `inline if / inline for / inline while` but so far no breaking changes were needed and a big portion of dotty syntax is already parsed.

dotty syntax still evolves so nobody can guarantee we won't have to again introduce breaking changes :shrug:

**Comment by domdorn** [comment]
I just tried scalafmt 2.7.5 on my new scala3 microservice and noticed its not able to handle package scoped code (not sure whats the exact name, but the "not needing package-objects anymore") like 
```

package com.acme.calculation.demo.web

import akka.http.scaladsl.server.Route
import com.acme.calculation.demo.web.controller.DemoController

trait WebModule {
  def routes: Route
}

type WebModuleService = zio.Has[WebModule] // <-- here it fails

object WebModule {
  lazy val demoController = new DemoController
  import akka.http.scaladsl.server.Directives._
  
  lazy val live: WebModule = new WebModule {
    override def routes: Route = pathPrefix("api" / "v1" ) {
      demoController.route
    }
  }
  
}


```

Is there any update or snapshot I can already try?

**Comment by tgodzik** [comment]
@domdorn we haven't yet released a full version, but you can test it out via coursier:

https://scalameta.org/scalafmt/docs/installation.html#standalone

For example I used the most recent snapshot on Metals here:
https://github.com/scalameta/metals/pull/2587

You need to also add `runner.dialect = scala3` option to `scalafmt.conf`

**Comment by smarter** [comment]
To parse Scala 3 correctly there's another subtle difference that needs to be handled: infix type applications are parsed following precedence rules, like infix term applications, for example:
```scala
class ^[A, B]
trait Test {
  def foo: String Map Int ^ Double
}
```
In Scala 2, the type of `foo` is `(String Map Int) ^ Double` aka `^[Map[String, Int], Double]` but in Scala 3 it's `String Map (Int ^ Double)` aka `Map[String, ^[Int, Double]]`, there's an unmerged PR implementing this for Scala 2: https://github.com/scala/scala/pull/6147

**Comment by tgodzik** [comment]
> To parse Scala 3 correctly there's another subtle difference that needs to be handled: infix type applications are parsed following precedence rules, like infix term applications, for example:
> 
> ```scala
> class ^[A, B]
> trait Test {
>   def foo: String Map Int ^ Double
> }
> ```
> 
> In Scala 2, the type of `foo` is `(String Map Int) ^ Double` aka `^[Map[String, Int], Double]` but in Scala 3 it's `String Map (Int ^ Double)` aka `Map[String, ^[Int, Double]]`, there's an unmerged PR implementing this for Scala 2: [scala/scala#6147](https://github.com/scala/scala/pull/6147)

och, that's a fun one! Funnily enough this should not influence most of the existing tooling, so I will close this one and create a separate issue that we can follow up on in the future.

Thanks for pointing this out!

**Comment by tgodzik** [comment]
Opened https://github.com/scalameta/scalameta/issues/2315
